docs/guides/typescript.md

```diff
@@ -65,12 +63,188 @@ const lightMachine = createMachine<LightContext, LightEvent>({
 - Events which you send to the machine will be strongly typed, offering you much more confidence in the payload shapes you'll be receiving.
 
+## The VSCode Extension <Badge text="4.29+" />
+
+Using our [VS Code extension](https://marketplace.visualstudio.com/items?itemName=mattpocock.xstate-vscode), you can automatically generate intelligent typings for XState.
+
+Here's how you can get started:
+
+1. Download and install the [VS Code extension](https://marketplace.visualstudio.com/items?itemName=mattpocock.xstate-vscode).
+2. Open a new file and create a new machine, passing the schema attributes:
+
+```ts
+import { createMachine, t } from 'xstate';
+
+const machine = createMachine({
+  schema: {
+    context: t<{ value: string }>(),
+    events: t<{ type: 'FOO'; value: string } | { type: 'BAR' }>()
+  },
+  initial: 'a',
+  states: {
+    a: {
+      on: {
+        FOO: {
+          actions: 'consoleLogValue',
+          target: 'b'
+        }
+      }
+    },
+    b: {
+      entry: 'consoleLogValueAgain'
+    }
+  }
+});
+```
+
+3. Add `tsTypes: true` to the machine and save the file:
+
+```ts
+const machine = createMachine({
+  tsTypes: true,
+  schema: {
+    context: t<{ value: string }>(),
+    events: t<{ type: 'FOO'; value: string } | { type: 'BAR' }>()
+  },
+  initial: 'a',
+  states: {/* ... */}
+});
+```
+
+4. The extension should automatically add a generic to the machine:
+
+```ts
+const machine = createMachine({
+  tsTypes: {} as import('./filename.typegen').Typegen[0],
+  schema: {
+    context: t<{ value: string }>(),
+    events: t<{ type: 'FOO'; value: string } | { type: 'BAR' }>()
+  },
+  initial: 'a',
+  states: {/* ... */}
+});
+```
+
+5. Add a second parameter into the `createMachine` call - this is where you implement the actions, services, guards and delays for the machine.
+
+```ts
+const machine = createMachine(
+  {
+    // Config here
+  },
+  {
+    actions: {
+      consoleLogValue: (context, event) => {
+        // Wow! event is typed to { type: 'FOO' }
+        console.log(event.value);
+      },
+      consoleLogValueAgain: (context, event) => {
+        // Wow! event is typed to { type: 'FOO' }
+        console.log(event.value);
+      }
+    }
+  }
+);
+```
+
+You'll notice that the events in the options are _strongly typed to the events that cause the action to be triggered_. This is true for actions, guards, services and delays.
+
+You'll also notice that `state.matches`, `tags` and other parts of the machine are now type-safe.
+
+### How to get the most out of the VS Code extension
+
+#### Use named actions/guards/services
+
+Our recommendation with this approach is to mostly use named actions/guards/services, not inline ones.
+
+This is optimal:
+
+```ts
+createMachine(
+  {
+    entry: ['sayHello']
+  },
+  {
+    actions: {
+      sayHello: () => {
+        console.log('Hello!');
+      }
+    }
+  }
+);
+```
+
+This is useful, but less optimal:
```

Comment:
this will prevent loading it in the editor, right? so in a way - it's not that useful 🤣

Reply:
Iiiish, you'll still be able to load the inspector, and generate typegen from it.

Reply:
should we even allow generating typegen with generics? what would happen in such a situation? would it be pushed to the last place of the generics list? would it be inserted at the position of `tsTypes` property?

Reply:
It gets inserted at the position - then you get a red line at the generics saying 'refactor this to schema'